File: 한승우/9월 4주차/k진수에서 소수 개수 구하기.cpp
#include <string>
#include <algorithm>
#include <sstream>
#include <iostream>
typedef long long ll;

using namespace std;

string convert(int n, int k){
    string res = "";
    
    while(n > 0){
        res += to_string(n % k);
        n /= k;
    }
    
    reverse(res.begin(), res.end());
    
    return res;
}

bool is_prime(ll num){
    if(num < 2) return false;
    
    for(ll i = 2; i * i <= num; i++)
        if(num % i == 0) return false;
    return true;
}

int solution(int n, int k) {
    int answer = 0;
    string num = convert(n, k); //k진법으로 변환
    //cout << num << endl;
    
    string token;
    vector<string> tokens;
    stringstream ss(num);
    
    //0으로 분할
    while(getline(ss, token, '0')){
        if(!token.empty()) {
            tokens.push_back(token);
            //cout << token << endl;
        }
    }
    
    //각 숫자가 소수인지 검사
    for(string s: tokens){
        if(is_prime(stoll(s))) answer++;
    }
    
    return answer;
}

File: 송하연/9월4주차/k진수에서 소수 개수 구하기.java
class Solution {
    public int solution(int n, int k) {
        int answer = 0;
 
        String convert = Integer.toString(n,k); //진법 변환 메서드
        String parts[] = convert.split("0");
        
        for(String str: parts){
            if(str.equals("")) continue;
            
            //int형은 21억까지 저장 가능. 진법 변환하면 초과 가능성
            long num = Long.parseLong(str);
            if(isPrime(num)) answer++;    
        }
        return answer;
    }

    
    boolean isPrime(long num) {
        if(num==0 || num==1) return false; 
        
        //제곱근 활용하기
        long sqrt = (long)Math.sqrt(num);
        
        for(long i=2; i<=sqrt; i++){
            if(num%i==0) return false;
        }
        return true; 
    }
}
